sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +298,190 @@ def reduce(self, word, exclude=None):
+                    next_sub = next_sub.subword(1, len(next_sub))
+                # Check if the transition to any new state in posible.
+                for sub_word in current_state_name_array:
+                    if sub_word in accept_states:
+                        # Add accept states.
+                        if current_state_type == "accept":
+                            for letter in automaton_alphabet:
+                                next = current_state_name*letter
+                                while True:
+                                    if len(next) <= 1:
+                                        if next in self.reduction_automaton.states:
+                                            self.reduction_automaton.states[state].add_transition(letter, next)
+                                        else:
+                                            self.reduction_automaton.states[state].add_transition(letter, 'start')
+                                        break
+                                    else:
+                                        if next in self.reduction_automaton.states:
+                                            self.reduction_automaton.states[state].add_transition(letter, next)
+                                            break
+                                        next = next.subword(1, len(next))
+                        break
+
+        # Add transitions for new states. All symbols used in the automaton are considered here.
+        # Ignore this if `generators` = `automaton_alphabet`.
+        if len(generators) != len(automaton_alphabet):
+            for state in proper_prefixes:
+                for letter in generators:
+                    next = current_state_name*letter
+                    len_next_word = len(next)
+                    while True:
+                        if len(next) <= 1:
+                            if next in self.reduction_automaton.states:
+                                self.reduction_automaton.states[state].add_transition(letter, next)
+                            else:
+                                self.reduction_automaton.states[state].add_transition(letter, 'start')
+                            break
+                        else:
+                            if next in self.reduction_automaton.states:
+                                self.reduction_automaton.states[state].add_transition(letter, next)
+                                break
+                            next = next.subword(1, len(next))
+
+    def reduce_using_automaton(self, word):
+        '''
+        Reduce a word using an automaton.
+
+        Summary:
+        All the elements of the automaton are stored in an array and are given as the input to the automaton.
+        If the automaton reaches a dead state that subword is replaced and the automaton is run from the beginning.
+        This is repeated until the word reaches the end and the automaton stays in the accept state.
```

Comment:
until the automaton reaches the end of the word. What do you mean by "it stays in the accept state"?

Reply:
I should update this too. 

Basically, it means that by the time we scan every symbol in the word (or by the time we reach the end), the complete word has to be replaced(according to the corresponding rule) if we reach a dead state. So, we terminate only when the word is completely read and we reach an accept state.

Reply:
I see, this makes more sense.